### PR TITLE
Port the individual test script runner to D

### PR DESCRIPTION
**Pull request by wilzbach**
- The individual test script runner is still "in development" (it has been added two weeks ago in https://github.com/dlang/dmd/pull/8040)
- As this is the "main repo" of the D programming language, we should be able to write our scripts in D too
- `rdmd` caches the build and subsequent executions don't require a compilation step of this tool (so there shouldn't be a noticeable slowdown in comparison to bash)
- The long-term goal is to skip the Makefile middleman, s.t. executing the testsuite on Windows doesn't depend on GNUmake

See also: #8129 (for another step on the journey from GNUMake to D)

### CODE REVIEW

File: test/run_individual_tests.d [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/**
+Allows running tests individually
+
+Usage:
+  ./run_individual_tests.d <test-file>...
+
+Example:
+  ./run_individual_tests.d runnable/template2962.d fail_compilation/fail282.d
+
+See the README.md for all available test targets
```
Comment (wilzbach) [inline test/run_individual_tests.d]: Maybe drop this header in favor of the built-in help and README.md?

File: test/run_individual_tests.d [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    string[] targets;
+    foreach (file; args[1 .. $])
+    {
+        auto normalizedDir = file.absolutePath.dirName;
+        targets ~= "test_results/%s/%s.out".format(normalizedDir.baseName, file.baseName);
+    }
+    auto scriptDir = __FILE_FULL_PATH__.dirName;
+    auto makeArguments = targets;
+    makeArguments ~= "--jobs=".text(jobs);
+    spawnProcess("make" ~ makeArguments, null, Config.none, scriptDir).wait;
```
Comment (wilzbach) [inline test/run_individual_tests.d]: With #8129, all that is left to replace the Makefile is:
- only add targets which have a target with an older or non-existing timestamp
- replace jobs with `foreach (target; targets.parallel(1))`
- ensure the `test_results` directories exists
- set the environment flags
- resolve shorthands like `all` or `run_fail_compilation_tests`
- ensure that the test sanitizers are build
- add `if (targets.canFind("clean") { return removeRecursive(targetDir);}`

File: test/run_individual_tests.d [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+`, res.options);
+        return;
+    }
+
+    string[] targets;
+    foreach (file; args[1 .. $])
+    {
+        auto normalizedDir = file.absolutePath.dirName;
+        targets ~= "test_results/%s/%s.out".format(normalizedDir.baseName, file.baseName);
+    }
```
Comment (jacob-carlborg) [inline test/run_individual_tests.d]: Replace with `map`?

File: test/run_individual_tests.d [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    auto scriptDir = __FILE_FULL_PATH__.dirName;
+    auto makeArguments = targets;
```
Comment (jacob-carlborg) [inline test/run_individual_tests.d]: Why the extra variable?

File: test/run_individual_tests.d [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    makeArguments ~= "--jobs=".text(jobs);
```
Comment (jacob-carlborg) [inline test/run_individual_tests.d]: Perhaps concatenate this on the previous line.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: test/run_individual_tests.d (hunk 3 above)
Comment (wilzbach) [reply test/run_individual_tests.d]: Ok. Done.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by wilzbach** [comment]
> So this supposed to be a shorthand form of running dcompiler -flags test1234.d yourself? :-)

Yes, but do you have _auto-completion_ for that? :stuck_out_tongue: 

> but the lengths that some people will go to in order to avoid typing...

This actually originates from the shell scripts which you can't easily run with `dcompiler -flags`.
Also there are many tests which do more than just compiling a file (ddoc, json, imports, other postprocessing) which you don't want to do by hand.
Other goodies:
- Run all diagnostics tests -> `./run_individual_tests.d ./fail_compilation/diag*`
- full features of the test suite like e.g. setting up all environment variables, diffing the output, maybe [auto-displaying a stack trace soon](https://github.com/dlang/dmd/pull/8132) ...
- a step towards replacing the Makefile which apparently ([1], [2], [3], [4]) is a huge problem for people on Windows

See https://github.com/dlang/dmd/pull/8040 for more details/motivation.

[1]  https://forum.dlang.org/post/tulbhulbeqqzofdxevcg@forum.dlang.org
[2] https://forum.dlang.org/post/gxxfmrnezfrlodlhpiwe@forum.dlang.org
[3] https://forum.dlang.org/post/uyzscudefkhqyjdzvczc@forum.dlang.org
[4] https://issues.dlang.org/show_bug.cgi?id=18413

**Review by JinShil — APPROVED** [review]

**Pull request merged**